File: python/1.py
from pathlib import Path

target = 2020

text = Path("../input1.txt").read_text()

numbers = []
for line in text.split("\n"):
    if not line:
        continue
    
    number = int(line)
    if number >= target:
        continue

    numbers.append(number)

numbers = sorted(numbers)

def f(numbers, target):
    for i in numbers:
        for j in reversed(numbers):
            s = i + j
            if s < target:
                break

            if s == target:
                return i, j

    return None


def part1():
    result = f(numbers, target)
    if result is None:
        print("No solution")
    else:
        i, j = result
        print(i * j)


def part2():
    for i, number in enumerate(numbers):
        result = f(numbers[i + 1:], target - number)
        if result:
            x, y = result
            print(x * y * number)
            return

    print("No solution")


part1()
part2()

File: rust/1.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::path::Path;

fn solve(numbers: &Vec<i32>, target: i32) -> Option<i32> {
    for i in numbers.iter() {
        for j in numbers.iter().rev() {
            let s = i + j;
            if s < target {
                break;
            }

            if s == target {
                return Some(i * j);
            }
        }
    }

    return None;
}

fn part1(numbers: &Vec<i32>, target: i32) {
    match solve(&numbers, target) {
        Some(result) => println!("{}", result),
        None => println!("No solution"),
    };
}

fn part2(numbers: &Vec<i32>, target: i32) {
    for (i, number) in numbers.iter().enumerate() {
        let new_target = target - number;
        match solve(&numbers[i + 1..].to_vec(), new_target) {
            Some(result) => println!("{}", number * result),
            None => (),
        };
    }
}

fn main() {
    let path = Path::new("../input1.txt");

    let file = match File::open(&path) {
        Err(why) => panic!("Couldn't open {}: {}", path.display(), why),
        Ok(file) => file,
    };

    let reader = BufReader::new(file);
    let numbers = {
        let numbers_iter = reader
            .lines()
            .map(|line| line.unwrap().parse::<i32>().unwrap());
        let mut numbers = numbers_iter.collect::<Vec<i32>>();
        numbers.sort();
        numbers
    };

    let target = 2020;
    part1(&numbers, target);
    part2(&numbers, target);
}
